ase/api/src/buffer.c

```diff
@@ -26,122 +26,402 @@
 
 #ifdef HAVE_CONFIG_H
 #include <config.h>
-#endif				// HAVE_CONFIG_H
+#endif // HAVE_CONFIG_H
 
-#include <opae/access.h>
-#include <opae/utils.h>
+#include "opae/access.h"
+#include "opae/utils.h"
 #include "common_int.h"
-#include <ase_common.h>
-
-#include <errno.h>
-#include <malloc.h>		/* malloc */
-#include <stdlib.h>		/* exit */
-#include <stdio.h>		/* printf */
-#include <string.h>		/* memcpy */
-#include <unistd.h>		/* getpid */
-#include <sys/types.h>		/* pid_t */
-#include <sys/ioctl.h>		/* ioctl */
-#include <sys/mman.h>		/* mmap & munmap */
-#include <sys/time.h>		/* struct timeval */
-
-
-fpga_result __FPGA_API__ fpgaPrepareBuffer(fpga_handle handle,
-					   uint64_t len, void **buf_addr,
-					   uint64_t *wsid, int flags)
+
+#include "ase_common.h"
+#include "ase_host_memory.h"
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <unistd.h>
+
+/* Others */
+#define KB 1024
+#define MB (1024 * KB)
+#define GB (1024UL * MB)
+
+#define PROTECTION (PROT_READ | PROT_WRITE)
+
+#ifndef MAP_HUGETLB
+#define MAP_HUGETLB 0x40000
+#endif
+#ifndef MAP_HUGE_SHIFT
+#define MAP_HUGE_SHIFT 26
+#endif
+
+#define MAP_1G_HUGEPAGE	(0x1e << MAP_HUGE_SHIFT) /* 2 ^ 0x1e = 1G */
+
+#ifdef __ia64__
+#define ADDR (void *)(0x8000000000000000UL)
+#define FLAGS_4K (MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED)
+#define FLAGS_2M (FLAGS_4K | MAP_HUGETLB)
+#define FLAGS_1G (FLAGS_2M | MAP_1G_HUGEPAGE)
+#else
+#define ADDR (void *)(0x0UL)
+#define FLAGS_4K (MAP_PRIVATE | MAP_ANONYMOUS)
+#define FLAGS_2M (FLAGS_4K | MAP_HUGETLB)
+#define FLAGS_1G (FLAGS_2M | MAP_1G_HUGEPAGE)
+#endif
+
+
+/*
+ * In simulation the physical page assignment is synthetic. We can afford to
```

Comment:
comments are very helpful

Reply:
I agree comments are helpful, but buffer.c here matches the OPAE buffer.c as closely as possible. My goal when adding files in the api directory was to be as close as possible to OPAE. Some day I'd like to share code between OPAE and ASE and have simulator/hardware functions separated out.

I think you'll find the ASE-specific changes have better comments.